src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -501,43 +530,63 @@ def expectation_from_circuit(self, circuit: "Circuit", nshots: int = 1000) -> fl
+            measurements = {}
+            for term in terms:
+                # store coefficient
+                tmp_coeffs.append(term.coefficient)
+
+                # Only care about non-I terms
+                non_identity_factors = []
+                # prepare the measurement basis and append it to the circuit
+                for factor in term.factors:
+                    if factor.name[0] != "I":
+                        non_identity_factors.append(factor)
+                        q = factor.target_qubit
+                        if q not in measurements:
+                            measurements[q] = gates.M(q, basis=factor.gate.__class__)
+
+                # build diagonal observable
+                tmp_obs.append(
+                    SymbolicHamiltonian(
+                        prod(Z(factor.target_qubit) for factor in non_identity_factors),
+                        nqubits=circuit.nqubits,
+                        backend=self.backend,
+                    )
                 )
-            )
-            # Get the qubits we want to measure for each term
-            qubit_map = sorted(factor.target_qubit for factor in non_identity_factors)
-            # prepare the measurement basis and append it to the circuit
-            measurements = [
-                gates.M(factor.target_qubit, basis=factor.gate.__class__)
-                for factor in non_identity_factors
-            ]
+
+                # Get the qubits we want to measure for each term
+                tmp_qmaps.append(
+                    sorted(factor.target_qubit for factor in non_identity_factors)
+                )
+            coefficients.append(tmp_coeffs)
+            Z_observables.append(tmp_obs)
+            qubit_maps.append(tmp_qmaps)
+
             circ_copy = circuit.copy(True)
-            circ_copy.add(measurements)
+            circ_copy.add(list(measurements.values()))
             rotated_circuits.append(circ_copy)
-            # for mapping the obtained sample frequencies to the original qubits
-            qubit_maps.append(qubit_map)
-        frequencies = [
-            result.frequencies()
-            for result in self.backend.execute_circuits(rotated_circuits, nshots=nshots)
-        ]
-        return sum(
-            coeff * obs.expectation_from_samples(freq, qubit_map)
-            for coeff, freq, obs, qubit_map in zip(
-                coefficients, frequencies, Z_observables, qubit_maps
+
+        # execute the circuits
+        results = self.backend.execute_circuits(rotated_circuits, nshots=nshots)
+
+        # construct the expectation value for each diagonal term
+        # and sum all together
+        expval = 0.0
+        for coeffs, res, obs, qmaps in zip(
+            coefficients, results, Z_observables, qubit_maps
+        ):
+            freq = res.frequencies()
+            expval += sum(
+                c.real * o.expectation_from_samples(freq, qmap)
+                for c, o, qmap in zip(coeffs, obs, qmaps)
```

Comment:
```suggestion
            expval += sum(
                c.real * o.expectation_from_samples(freq, qmaps)
                for c, o in zip(coeffs, obs)
```

- Changing to use the same `qubit_map` for all the terms that can be measured simultaneously
- Should be able to take out `c.real` too if you decide to put `term.coefficient` inside the diagonal Hamiltonian